# Super-Resolution for Strong Lensing Images

This notebook implements a deep learning-based super-resolution approach for strong lensing images. It is divided into two main tasks:

## Task III.A: Super-Resolution on Simulated Data

**Objective:**  
Upscale low-resolution simulated strong lensing images using high-resolution samples as ground truths.

**Dataset:**  
- **Link:** [Task III.A](https://drive.google.com/file/d/1uJmDZw649XS-r-dYs9WD-OPwF_TIroVw/view?usp=sharing)

**Implementation Steps:**
1. **Data Loading and Preprocessing:**  
   - Used a custom PyTorch `Dataset` class to load `.npy` files for HR and LR images.
   - Created a 90:10 train–validation split.
2. **Model Architecture:**  
   - Implementd a baseline model (e.g., SRCNN) that first upsamples the LR image (using bilinear interpolation) and then refines it with convolutional layers.
3. **Training Loop:**  
   - Traind the model using MSE loss.
   - Tracked training and validation metrics (MSE, PSNR, SSIM).
   - Used early stopping and saved the best model.
4. **Evaluation:**  
   - Computed evaluation metrics on the validation set.
   - Displayed sample outputs for qualitative evaluation.

In [28]:
import os
import numpy as np
from torch.utils.data import Dataset, DataLoader, random_split
import torch
import math
import torch.nn as nn
import torch.optim as optim
from skimage.metrics import peak_signal_noise_ratio, mean_squared_error, structural_similarity
from tqdm import tqdm 
import torchvision.transforms as transforms

In [9]:
class SuperResolutionDataset(Dataset):
    """
    A PyTorch Dataset class for super-resolution tasks.
    It expects two directories: one for high-resolution (HR) images and one for low-resolution (LR) images.
    The images are stored as .npy files.
    """
    def __init__(self, hr_dir, lr_dir, transform=None):
        self.hr_dir = hr_dir
        self.lr_dir = lr_dir
        self.hr_files = sorted([os.path.join(hr_dir, f) for f in os.listdir(hr_dir) if f.endswith('.npy')])
        self.lr_files = sorted([os.path.join(lr_dir, f) for f in os.listdir(lr_dir) if f.endswith('.npy')])
        self.transform = transform

        # Ensure both lists have the same length
        assert len(self.hr_files) == len(self.lr_files), "Mismatch between HR and LR files count!"

    def __len__(self):
        return len(self.hr_files)

    def __getitem__(self, idx):
        hr = np.load(self.hr_files[idx])
        lr = np.load(self.lr_files[idx])
        
        # Convert to float32 and add channel dimension if needed
        # (Assuming images are grayscale. Modify if they are multi-channel)
        hr = hr.astype(np.float32)
        lr = lr.astype(np.float32)
        if hr.ndim == 2:
            hr = hr[None, :, :]  # Add channel dimension
        if lr.ndim == 2:
            lr = lr[None, :, :]
        
        sample = {'lr': torch.from_numpy(lr), 'hr': torch.from_numpy(hr)}
        
        if self.transform:
            sample = self.transform(sample)
            
        return sample

In [12]:
task3a_hr_dir = '/kaggle/input/task3a/Dataset/HR'
task3a_lr_dir = '/kaggle/input/task3a/Dataset/LR'

# Create the dataset instance (already tested earlier)
dataset_task3a = SuperResolutionDataset(task3a_hr_dir, task3a_lr_dir)

In [19]:
class SRCNN(nn.Module):
    def __init__(self, num_channels=1):
        super(SRCNN, self).__init__()
        self.conv1 = nn.Conv2d(num_channels, 64, kernel_size=9, padding=4)
        self.conv2 = nn.Conv2d(64, 32, kernel_size=5, padding=2)
        self.conv3 = nn.Conv2d(32, num_channels, kernel_size=5, padding=2)
        self.relu = nn.ReLU(inplace=True)
    
    def forward(self, x):
        x = self.relu(self.conv1(x))
        x = self.relu(self.conv2(x))
        x = self.conv3(x)
        return x

In [20]:
class SuperResolutionModel(nn.Module):
    def __init__(self, scale_factor=2):
        super(SuperResolutionModel, self).__init__()
        self.scale_factor = scale_factor
        self.srcnn = SRCNN(num_channels=1)
        
    def forward(self, x):
        # Upsample LR image to HR size using bilinear interpolation
        x_up = nn.functional.interpolate(x, scale_factor=self.scale_factor, mode='bilinear', align_corners=False)
        out = self.srcnn(x_up)
        return out

In [21]:
task3a_hr_dir = '/kaggle/input/task3a/Dataset/HR'
task3a_lr_dir = '/kaggle/input/task3a/Dataset/LR'

# Assume SuperResolutionDataset is already defined as in previous steps
dataset_task3a = SuperResolutionDataset(task3a_hr_dir, task3a_lr_dir)
train_size = int(0.9 * len(dataset_task3a))
val_size = len(dataset_task3a) - train_size
train_dataset, val_dataset = random_split(dataset_task3a, [train_size, val_size])
train_loader = DataLoader(train_dataset, batch_size=16, shuffle=True, num_workers=2)
val_loader = DataLoader(val_dataset, batch_size=16, shuffle=False, num_workers=2)

In [22]:
model = SuperResolutionModel(scale_factor=2).cuda()
criterion = nn.MSELoss()
optimizer = optim.Adam(model.parameters(), lr=1e-4)

In [23]:
class EarlyStopping:
    """
    Early stops the training if validation loss doesn't improve after a given patience.
    """
    def __init__(self, patience=5, verbose=False):
        self.patience = patience
        self.verbose = verbose
        self.counter = 0
        self.best_loss = None
        self.early_stop = False

    def __call__(self, val_loss):
        if self.best_loss is None or val_loss < self.best_loss:
            if self.verbose and self.best_loss is not None:
                print(f"Validation loss improved from {self.best_loss:.6f} to {val_loss:.6f}.")
            self.best_loss = val_loss
            self.counter = 0
        else:
            self.counter += 1
            if self.verbose:
                print(f"Validation loss did not improve. EarlyStopping counter: {self.counter}/{self.patience}")
            if self.counter >= self.patience:
                self.early_stop = True

In [24]:
def evaluate(model, dataloader):
    model.eval()
    total_mse = 0.0
    total_psnr = 0.0
    total_ssim = 0.0
    count = 0
    with torch.no_grad():
        for batch in dataloader:
            lr = batch['lr'].cuda()
            hr = batch['hr'].cuda()
            sr = model(lr)
            for i in range(sr.size(0)):
                sr_img = sr[i].cpu().numpy().squeeze()
                hr_img = hr[i].cpu().numpy().squeeze()
                mse_val = mean_squared_error(hr_img, sr_img)
                psnr_val = peak_signal_noise_ratio(hr_img, sr_img, data_range=hr_img.max()-hr_img.min())
                ssim_val = structural_similarity(hr_img, sr_img, data_range=hr_img.max()-hr_img.min())
                total_mse += mse_val
                total_psnr += psnr_val
                total_ssim += ssim_val
                count += 1
    avg_mse = total_mse / count
    avg_psnr = total_psnr / count
    avg_ssim = total_ssim / count
    return avg_mse, avg_psnr, avg_ssim

In [27]:
num_epochs = 50  # Maximum number of epochs
early_stopping = EarlyStopping(patience=5, verbose=True)
best_val_loss = float('inf')  # To track the best validation loss

print("Starting training...\n")
for epoch in range(num_epochs):
    model.train()
    running_loss = 0.0
    train_bar = tqdm(train_loader, desc=f"Epoch {epoch+1}/{num_epochs} [Training]", leave=False)
    for batch in train_bar:
        lr = batch['lr'].cuda()
        hr = batch['hr'].cuda()
        optimizer.zero_grad()
        sr = model(lr)
        loss = criterion(sr, hr)
        loss.backward()
        optimizer.step()
        running_loss += loss.item() * lr.size(0)
        train_bar.set_postfix(loss=loss.item())
    
    train_loss = running_loss / len(train_loader.dataset)
    
    # Validation loop with progress bar
    model.eval()
    val_running_loss = 0.0
    with torch.no_grad():
        val_bar = tqdm(val_loader, desc=f"Epoch {epoch+1}/{num_epochs} [Validation]", leave=False)
        for batch in val_bar:
            lr = batch['lr'].cuda()
            hr = batch['hr'].cuda()
            sr = model(lr)
            loss = criterion(sr, hr)
            val_running_loss += loss.item() * lr.size(0)
            val_bar.set_postfix(loss=loss.item())
    val_loss = val_running_loss / len(val_loader.dataset)
    
    # Compute additional metrics for validation set
    val_mse, val_psnr, val_ssim = evaluate(model, val_loader)
    
    print(f"Epoch {epoch+1}/{num_epochs} | Train Loss: {train_loss:.6f} | Val Loss: {val_loss:.6f} | "
          f"Val MSE: {val_mse:.6f} | PSNR: {val_psnr:.2f} | SSIM: {val_ssim:.4f}")
    
    # Save the best model
    if val_loss < best_val_loss:
        best_val_loss = val_loss
        torch.save(model.state_dict(), 'best_model.pth')
        print(f"--> Best model saved at epoch {epoch+1} with validation loss: {val_loss:.6f}")
    
    # Check for early stopping
    early_stopping(val_loss)
    if early_stopping.early_stop:
        print("Early stopping triggered. Ending training.")
        break

print("Training complete. Best model saved as 'best_model.pth'.")

Starting training...



Epoch 1/50 | Train Loss: 0.000232 | Val Loss: 0.000079 | Val MSE: 0.000079 | PSNR: 41.04 | SSIM: 0.9694
--> Best model saved at epoch 1 with validation loss: 0.000079


Epoch 2/50 | Train Loss: 0.000073 | Val Loss: 0.000069 | Val MSE: 0.000069 | PSNR: 41.65 | SSIM: 0.9732
--> Best model saved at epoch 2 with validation loss: 0.000069
Validation loss improved from 0.000079 to 0.000069.


Epoch 3/50 | Train Loss: 0.000068 | Val Loss: 0.000067 | Val MSE: 0.000067 | PSNR: 41.80 | SSIM: 0.9741
--> Best model saved at epoch 3 with validation loss: 0.000067
Validation loss improved from 0.000069 to 0.000067.


Epoch 4/50 | Train Loss: 0.000066 | Val Loss: 0.000066 | Val MSE: 0.000066 | PSNR: 41.87 | SSIM: 0.9746
--> Best model saved at epoch 4 with validation loss: 0.000066
Validation loss improved from 0.000067 to 0.000066.


Epoch 5/50 | Train Loss: 0.000065 | Val Loss: 0.000065 | Val MSE: 0.000065 | PSNR: 41.92 | SSIM: 0.9750
--> Best model saved at epoch 5 with validation loss: 0.000065
Validation loss improved from 0.000066 to 0.000065.


Epoch 6/50 | Train Loss: 0.000065 | Val Loss: 0.000064 | Val MSE: 0.000064 | PSNR: 41.95 | SSIM: 0.9752
--> Best model saved at epoch 6 with validation loss: 0.000064
Validation loss improved from 0.000065 to 0.000064.


Epoch 7/50 | Train Loss: 0.000064 | Val Loss: 0.000065 | Val MSE: 0.000065 | PSNR: 41.91 | SSIM: 0.9752
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 8/50 | Train Loss: 0.000064 | Val Loss: 0.000066 | Val MSE: 0.000066 | PSNR: 41.86 | SSIM: 0.9749
Validation loss did not improve. EarlyStopping counter: 2/5


Epoch 9/50 | Train Loss: 0.000064 | Val Loss: 0.000063 | Val MSE: 0.000063 | PSNR: 42.02 | SSIM: 0.9756
--> Best model saved at epoch 9 with validation loss: 0.000063
Validation loss improved from 0.000064 to 0.000063.


Epoch 10/50 | Train Loss: 0.000064 | Val Loss: 0.000063 | Val MSE: 0.000063 | PSNR: 42.03 | SSIM: 0.9757
--> Best model saved at epoch 10 with validation loss: 0.000063
Validation loss improved from 0.000063 to 0.000063.


Epoch 11/50 | Train Loss: 0.000063 | Val Loss: 0.000063 | Val MSE: 0.000063 | PSNR: 42.06 | SSIM: 0.9758
--> Best model saved at epoch 11 with validation loss: 0.000063
Validation loss improved from 0.000063 to 0.000063.


Epoch 12/50 | Train Loss: 0.000063 | Val Loss: 0.000063 | Val MSE: 0.000063 | PSNR: 42.07 | SSIM: 0.9759
--> Best model saved at epoch 12 with validation loss: 0.000063
Validation loss improved from 0.000063 to 0.000063.


Epoch 13/50 | Train Loss: 0.000063 | Val Loss: 0.000063 | Val MSE: 0.000063 | PSNR: 42.08 | SSIM: 0.9760
--> Best model saved at epoch 13 with validation loss: 0.000063
Validation loss improved from 0.000063 to 0.000063.


Epoch 14/50 | Train Loss: 0.000063 | Val Loss: 0.000063 | Val MSE: 0.000063 | PSNR: 42.02 | SSIM: 0.9756
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 15/50 | Train Loss: 0.000063 | Val Loss: 0.000062 | Val MSE: 0.000062 | PSNR: 42.10 | SSIM: 0.9761
--> Best model saved at epoch 15 with validation loss: 0.000062
Validation loss improved from 0.000063 to 0.000062.


Epoch 16/50 | Train Loss: 0.000062 | Val Loss: 0.000062 | Val MSE: 0.000062 | PSNR: 42.11 | SSIM: 0.9762
--> Best model saved at epoch 16 with validation loss: 0.000062
Validation loss improved from 0.000062 to 0.000062.


Epoch 17/50 | Train Loss: 0.000062 | Val Loss: 0.000062 | Val MSE: 0.000062 | PSNR: 42.11 | SSIM: 0.9762
--> Best model saved at epoch 17 with validation loss: 0.000062
Validation loss improved from 0.000062 to 0.000062.


Epoch 18/50 | Train Loss: 0.000062 | Val Loss: 0.000062 | Val MSE: 0.000062 | PSNR: 42.12 | SSIM: 0.9762
--> Best model saved at epoch 18 with validation loss: 0.000062
Validation loss improved from 0.000062 to 0.000062.


Epoch 19/50 | Train Loss: 0.000062 | Val Loss: 0.000063 | Val MSE: 0.000063 | PSNR: 42.07 | SSIM: 0.9762
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 20/50 | Train Loss: 0.000062 | Val Loss: 0.000062 | Val MSE: 0.000062 | PSNR: 42.12 | SSIM: 0.9762
--> Best model saved at epoch 20 with validation loss: 0.000062
Validation loss improved from 0.000062 to 0.000062.


Epoch 21/50 | Train Loss: 0.000062 | Val Loss: 0.000062 | Val MSE: 0.000062 | PSNR: 42.15 | SSIM: 0.9764
--> Best model saved at epoch 21 with validation loss: 0.000062
Validation loss improved from 0.000062 to 0.000062.


Epoch 22/50 | Train Loss: 0.000062 | Val Loss: 0.000061 | Val MSE: 0.000061 | PSNR: 42.15 | SSIM: 0.9764
--> Best model saved at epoch 22 with validation loss: 0.000061
Validation loss improved from 0.000062 to 0.000061.


Epoch 23/50 | Train Loss: 0.000062 | Val Loss: 0.000061 | Val MSE: 0.000061 | PSNR: 42.16 | SSIM: 0.9765
--> Best model saved at epoch 23 with validation loss: 0.000061
Validation loss improved from 0.000061 to 0.000061.


Epoch 24/50 | Train Loss: 0.000062 | Val Loss: 0.000061 | Val MSE: 0.000061 | PSNR: 42.17 | SSIM: 0.9765
--> Best model saved at epoch 24 with validation loss: 0.000061
Validation loss improved from 0.000061 to 0.000061.


Epoch 25/50 | Train Loss: 0.000061 | Val Loss: 0.000061 | Val MSE: 0.000061 | PSNR: 42.17 | SSIM: 0.9766
--> Best model saved at epoch 25 with validation loss: 0.000061
Validation loss improved from 0.000061 to 0.000061.


Epoch 26/50 | Train Loss: 0.000061 | Val Loss: 0.000061 | Val MSE: 0.000061 | PSNR: 42.18 | SSIM: 0.9766
--> Best model saved at epoch 26 with validation loss: 0.000061
Validation loss improved from 0.000061 to 0.000061.


Epoch 27/50 | Train Loss: 0.000061 | Val Loss: 0.000062 | Val MSE: 0.000062 | PSNR: 42.13 | SSIM: 0.9763
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 28/50 | Train Loss: 0.000061 | Val Loss: 0.000061 | Val MSE: 0.000061 | PSNR: 42.20 | SSIM: 0.9767
--> Best model saved at epoch 28 with validation loss: 0.000061
Validation loss improved from 0.000061 to 0.000061.


Epoch 29/50 | Train Loss: 0.000061 | Val Loss: 0.000061 | Val MSE: 0.000061 | PSNR: 42.20 | SSIM: 0.9767
--> Best model saved at epoch 29 with validation loss: 0.000061
Validation loss improved from 0.000061 to 0.000061.


Epoch 30/50 | Train Loss: 0.000061 | Val Loss: 0.000061 | Val MSE: 0.000061 | PSNR: 42.21 | SSIM: 0.9767
--> Best model saved at epoch 30 with validation loss: 0.000061
Validation loss improved from 0.000061 to 0.000061.


Epoch 31/50 | Train Loss: 0.000061 | Val Loss: 0.000061 | Val MSE: 0.000061 | PSNR: 42.21 | SSIM: 0.9768
--> Best model saved at epoch 31 with validation loss: 0.000061
Validation loss improved from 0.000061 to 0.000061.


Epoch 32/50 | Train Loss: 0.000061 | Val Loss: 0.000062 | Val MSE: 0.000062 | PSNR: 42.14 | SSIM: 0.9764
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 33/50 | Train Loss: 0.000061 | Val Loss: 0.000060 | Val MSE: 0.000060 | PSNR: 42.22 | SSIM: 0.9768
--> Best model saved at epoch 33 with validation loss: 0.000060
Validation loss improved from 0.000061 to 0.000060.


Epoch 34/50 | Train Loss: 0.000061 | Val Loss: 0.000060 | Val MSE: 0.000060 | PSNR: 42.23 | SSIM: 0.9768
--> Best model saved at epoch 34 with validation loss: 0.000060
Validation loss improved from 0.000060 to 0.000060.


Epoch 35/50 | Train Loss: 0.000061 | Val Loss: 0.000060 | Val MSE: 0.000060 | PSNR: 42.23 | SSIM: 0.9769
--> Best model saved at epoch 35 with validation loss: 0.000060
Validation loss improved from 0.000060 to 0.000060.


Epoch 36/50 | Train Loss: 0.000061 | Val Loss: 0.000060 | Val MSE: 0.000060 | PSNR: 42.23 | SSIM: 0.9769
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 37/50 | Train Loss: 0.000061 | Val Loss: 0.000060 | Val MSE: 0.000060 | PSNR: 42.23 | SSIM: 0.9769
--> Best model saved at epoch 37 with validation loss: 0.000060
Validation loss improved from 0.000060 to 0.000060.


Epoch 38/50 | Train Loss: 0.000061 | Val Loss: 0.000060 | Val MSE: 0.000060 | PSNR: 42.24 | SSIM: 0.9769
--> Best model saved at epoch 38 with validation loss: 0.000060
Validation loss improved from 0.000060 to 0.000060.


Epoch 39/50 | Train Loss: 0.000060 | Val Loss: 0.000060 | Val MSE: 0.000060 | PSNR: 42.24 | SSIM: 0.9769
--> Best model saved at epoch 39 with validation loss: 0.000060
Validation loss improved from 0.000060 to 0.000060.


Epoch 40/50 | Train Loss: 0.000060 | Val Loss: 0.000060 | Val MSE: 0.000060 | PSNR: 42.25 | SSIM: 0.9770
--> Best model saved at epoch 40 with validation loss: 0.000060
Validation loss improved from 0.000060 to 0.000060.


Epoch 41/50 | Train Loss: 0.000060 | Val Loss: 0.000060 | Val MSE: 0.000060 | PSNR: 42.25 | SSIM: 0.9770
--> Best model saved at epoch 41 with validation loss: 0.000060
Validation loss improved from 0.000060 to 0.000060.


Epoch 42/50 | Train Loss: 0.000060 | Val Loss: 0.000060 | Val MSE: 0.000060 | PSNR: 42.25 | SSIM: 0.9770
--> Best model saved at epoch 42 with validation loss: 0.000060
Validation loss improved from 0.000060 to 0.000060.


Epoch 43/50 | Train Loss: 0.000060 | Val Loss: 0.000060 | Val MSE: 0.000060 | PSNR: 42.26 | SSIM: 0.9770
--> Best model saved at epoch 43 with validation loss: 0.000060
Validation loss improved from 0.000060 to 0.000060.


Epoch 44/50 | Train Loss: 0.000060 | Val Loss: 0.000060 | Val MSE: 0.000060 | PSNR: 42.26 | SSIM: 0.9770
--> Best model saved at epoch 44 with validation loss: 0.000060
Validation loss improved from 0.000060 to 0.000060.


Epoch 45/50 | Train Loss: 0.000060 | Val Loss: 0.000060 | Val MSE: 0.000060 | PSNR: 42.26 | SSIM: 0.9770
--> Best model saved at epoch 45 with validation loss: 0.000060
Validation loss improved from 0.000060 to 0.000060.


Epoch 46/50 | Train Loss: 0.000060 | Val Loss: 0.000060 | Val MSE: 0.000060 | PSNR: 42.26 | SSIM: 0.9770
--> Best model saved at epoch 46 with validation loss: 0.000060
Validation loss improved from 0.000060 to 0.000060.


Epoch 47/50 | Train Loss: 0.000060 | Val Loss: 0.000060 | Val MSE: 0.000060 | PSNR: 42.26 | SSIM: 0.9771
--> Best model saved at epoch 47 with validation loss: 0.000060
Validation loss improved from 0.000060 to 0.000060.


Epoch 48/50 | Train Loss: 0.000060 | Val Loss: 0.000060 | Val MSE: 0.000060 | PSNR: 42.26 | SSIM: 0.9771
--> Best model saved at epoch 48 with validation loss: 0.000060
Validation loss improved from 0.000060 to 0.000060.


Epoch 49/50 | Train Loss: 0.000060 | Val Loss: 0.000060 | Val MSE: 0.000060 | PSNR: 42.26 | SSIM: 0.9770
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 50/50 | Train Loss: 0.000060 | Val Loss: 0.000060 | Val MSE: 0.000060 | PSNR: 42.27 | SSIM: 0.9771
--> Best model saved at epoch 50 with validation loss: 0.000060
Validation loss improved from 0.000060 to 0.000060.
Training complete. Best model saved as 'best_model.pth'.


## Task III.B: Transfer Learning on Real Data

**Objective:**  
Enhance low-resolution strong lensing images using a limited dataset of real HR/LR pairs collected from HSC and HST telescopes.

**Dataset:**  
- **Link:** [Task III.B](https://drive.google.com/file/d/1plYfM-jFJT7TbTMVssuCCFvLzGdxMQ4h/view?usp=sharing)


**Implementation Steps:**
1. **Data Loading:**  
   - Reused the dataset class.
2. **Transfer Learning:**  
   - Loaded the pre-trained model from Task III.A.
   - Fine-tuned the model on the task III.B dataset using a lower learning rate.
3. **Training and Evaluation:**  
   - Used early stopping to prevent overfitting.
   - Monitored and saved the best model based on validation loss.
   - Computed evaluation metrics (MSE, PSNR, SSIM) on the validation set.
4. **Results:**  
   - Compared performance with and without data augmentation.

In [29]:
task3b_hr_dir = '/kaggle/input/specific-task-3b/Dataset 3B/Dataset/HR'
task3b_lr_dir = '/kaggle/input/specific-task-3b/Dataset 3B/Dataset/LR'

In [30]:
class AugmentedSuperResolutionDataset(SuperResolutionDataset):
    def __init__(self, hr_dir, lr_dir, transform=None):
        super().__init__(hr_dir, lr_dir, transform)
    
    def __getitem__(self, idx):
        sample = super().__getitem__(idx)
        # Example augmentation: Random horizontal flip (you can add more augmentations if desired)
        if np.random.rand() > 0.5:
            sample['lr'] = torch.flip(sample['lr'], dims=[2])
            sample['hr'] = torch.flip(sample['hr'], dims=[2])
        return sample

In [31]:
dataset_task3b = AugmentedSuperResolutionDataset(task3b_hr_dir, task3b_lr_dir)
train_size = int(0.9 * len(dataset_task3b))
val_size = len(dataset_task3b) - train_size
train_dataset, val_dataset = random_split(dataset_task3b, [train_size, val_size])
train_loader = DataLoader(train_dataset, batch_size=8, shuffle=True, num_workers=2)
val_loader = DataLoader(val_dataset, batch_size=8, shuffle=False, num_workers=2)

In [32]:
model_ft = SuperResolutionModel(scale_factor=2).cuda()
pretrained_path = 'best_model.pth'
if os.path.exists(pretrained_path):
    model_ft.load_state_dict(torch.load(pretrained_path))
    print("Pre-trained weights loaded from Task III.A.")
else:
    print("Pre-trained model not found. Please ensure 'best_model.pth' is available.")

Pre-trained weights loaded from Task III.A.


<ipython-input-32-6a1ac7f23a97>:4: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  model_ft.load_state_dict(torch.load(pretrained_path))


In [33]:
criterion = nn.MSELoss()
# Use a lower learning rate for fine-tuning
optimizer = optim.Adam(model_ft.parameters(), lr=1e-5)

In [34]:
num_epochs = 50  # maximum number of epochs for fine-tuning
early_stopping = EarlyStopping(patience=5, verbose=True)
best_val_loss = float('inf')

print("Starting fine-tuning on Task III.B dataset...\n")
for epoch in range(num_epochs):
    model_ft.train()
    running_loss = 0.0
    train_bar = tqdm(train_loader, desc=f"Epoch {epoch+1}/{num_epochs} [Training]", leave=False)
    for batch in train_bar:
        lr = batch['lr'].cuda()
        hr = batch['hr'].cuda()
        optimizer.zero_grad()
        sr = model_ft(lr)
        loss = criterion(sr, hr)
        loss.backward()
        optimizer.step()
        running_loss += loss.item() * lr.size(0)
        train_bar.set_postfix(loss=loss.item())
    
    train_loss = running_loss / len(train_loader.dataset)
    
    # Validation loop
    model_ft.eval()
    val_running_loss = 0.0
    with torch.no_grad():
        val_bar = tqdm(val_loader, desc=f"Epoch {epoch+1}/{num_epochs} [Validation]", leave=False)
        for batch in val_bar:
            lr = batch['lr'].cuda()
            hr = batch['hr'].cuda()
            sr = model_ft(lr)
            loss = criterion(sr, hr)
            val_running_loss += loss.item() * lr.size(0)
            val_bar.set_postfix(loss=loss.item())
    val_loss = val_running_loss / len(val_loader.dataset)
    
    # Compute additional metrics
    val_mse, val_psnr, val_ssim = evaluate(model_ft, val_loader)
    
    print(f"Epoch {epoch+1}/{num_epochs} | Train Loss: {train_loss:.6f} | Val Loss: {val_loss:.6f} | "
          f"Val MSE: {val_mse:.6f} | PSNR: {val_psnr:.2f} | SSIM: {val_ssim:.4f}")
    
    # Save best model for Task III.B fine-tuning
    if val_loss < best_val_loss:
        best_val_loss = val_loss
        torch.save(model_ft.state_dict(), 'best_model_task3b.pth')
        print(f"--> Best model (Task III.B) saved at epoch {epoch+1} with validation loss: {val_loss:.6f}")
    
    early_stopping(val_loss)
    if early_stopping.early_stop:
        print("Early stopping triggered. Ending fine-tuning.")
        break

print("Fine-tuning complete. Best model saved as 'best_model_task3b.pth'.")

Starting fine-tuning on Task III.B dataset...



Epoch 1/50 | Train Loss: 0.006301 | Val Loss: 0.003610 | Val MSE: 0.003615 | PSNR: 26.68 | SSIM: 0.5073
--> Best model (Task III.B) saved at epoch 1 with validation loss: 0.003610


Epoch 2/50 | Train Loss: 0.002900 | Val Loss: 0.001485 | Val MSE: 0.001490 | PSNR: 30.44 | SSIM: 0.6207
--> Best model (Task III.B) saved at epoch 2 with validation loss: 0.001485
Validation loss improved from 0.003610 to 0.001485.


Epoch 3/50 | Train Loss: 0.001964 | Val Loss: 0.000959 | Val MSE: 0.000963 | PSNR: 31.77 | SSIM: 0.6468
--> Best model (Task III.B) saved at epoch 3 with validation loss: 0.000959
Validation loss improved from 0.001485 to 0.000959.


Epoch 4/50 | Train Loss: 0.001818 | Val Loss: 0.000854 | Val MSE: 0.000854 | PSNR: 32.23 | SSIM: 0.6902
--> Best model (Task III.B) saved at epoch 4 with validation loss: 0.000854
Validation loss improved from 0.000959 to 0.000854.


Epoch 5/50 | Train Loss: 0.001726 | Val Loss: 0.000794 | Val MSE: 0.000794 | PSNR: 32.51 | SSIM: 0.7255
--> Best model (Task III.B) saved at epoch 5 with validation loss: 0.000794
Validation loss improved from 0.000854 to 0.000794.


Epoch 6/50 | Train Loss: 0.001676 | Val Loss: 0.000779 | Val MSE: 0.000779 | PSNR: 32.56 | SSIM: 0.7409
--> Best model (Task III.B) saved at epoch 6 with validation loss: 0.000779
Validation loss improved from 0.000794 to 0.000779.


Epoch 7/50 | Train Loss: 0.001644 | Val Loss: 0.000752 | Val MSE: 0.000752 | PSNR: 32.63 | SSIM: 0.7521
--> Best model (Task III.B) saved at epoch 7 with validation loss: 0.000752
Validation loss improved from 0.000779 to 0.000752.


Epoch 8/50 | Train Loss: 0.001626 | Val Loss: 0.000733 | Val MSE: 0.000736 | PSNR: 32.66 | SSIM: 0.7583
--> Best model (Task III.B) saved at epoch 8 with validation loss: 0.000733
Validation loss improved from 0.000752 to 0.000733.


Epoch 9/50 | Train Loss: 0.001603 | Val Loss: 0.000718 | Val MSE: 0.000717 | PSNR: 32.74 | SSIM: 0.7642
--> Best model (Task III.B) saved at epoch 9 with validation loss: 0.000718
Validation loss improved from 0.000733 to 0.000718.


Epoch 10/50 | Train Loss: 0.001589 | Val Loss: 0.000719 | Val MSE: 0.000718 | PSNR: 32.74 | SSIM: 0.7650
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 11/50 | Train Loss: 0.001577 | Val Loss: 0.000704 | Val MSE: 0.000701 | PSNR: 32.81 | SSIM: 0.7696
--> Best model (Task III.B) saved at epoch 11 with validation loss: 0.000704
Validation loss improved from 0.000718 to 0.000704.


Epoch 12/50 | Train Loss: 0.001577 | Val Loss: 0.000698 | Val MSE: 0.000694 | PSNR: 32.88 | SSIM: 0.7715
--> Best model (Task III.B) saved at epoch 12 with validation loss: 0.000698
Validation loss improved from 0.000704 to 0.000698.


Epoch 13/50 | Train Loss: 0.001563 | Val Loss: 0.000676 | Val MSE: 0.000677 | PSNR: 32.91 | SSIM: 0.7762
--> Best model (Task III.B) saved at epoch 13 with validation loss: 0.000676
Validation loss improved from 0.000698 to 0.000676.


Epoch 14/50 | Train Loss: 0.001544 | Val Loss: 0.000672 | Val MSE: 0.000670 | PSNR: 32.98 | SSIM: 0.7776
--> Best model (Task III.B) saved at epoch 14 with validation loss: 0.000672
Validation loss improved from 0.000676 to 0.000672.


Epoch 15/50 | Train Loss: 0.001527 | Val Loss: 0.000633 | Val MSE: 0.000631 | PSNR: 33.16 | SSIM: 0.7919
--> Best model (Task III.B) saved at epoch 15 with validation loss: 0.000633
Validation loss improved from 0.000672 to 0.000633.


Epoch 16/50 | Train Loss: 0.001519 | Val Loss: 0.000644 | Val MSE: 0.000645 | PSNR: 33.09 | SSIM: 0.7875
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 17/50 | Train Loss: 0.001518 | Val Loss: 0.000600 | Val MSE: 0.000599 | PSNR: 33.33 | SSIM: 0.8036
--> Best model (Task III.B) saved at epoch 17 with validation loss: 0.000600
Validation loss improved from 0.000633 to 0.000600.


Epoch 18/50 | Train Loss: 0.001508 | Val Loss: 0.000622 | Val MSE: 0.000622 | PSNR: 33.22 | SSIM: 0.7937
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 19/50 | Train Loss: 0.001492 | Val Loss: 0.000616 | Val MSE: 0.000616 | PSNR: 33.25 | SSIM: 0.7949
Validation loss did not improve. EarlyStopping counter: 2/5


Epoch 20/50 | Train Loss: 0.001491 | Val Loss: 0.000634 | Val MSE: 0.000632 | PSNR: 33.18 | SSIM: 0.7889
Validation loss did not improve. EarlyStopping counter: 3/5


Epoch 21/50 | Train Loss: 0.001472 | Val Loss: 0.000585 | Val MSE: 0.000588 | PSNR: 33.43 | SSIM: 0.8050
--> Best model (Task III.B) saved at epoch 21 with validation loss: 0.000585
Validation loss improved from 0.000600 to 0.000585.


Epoch 22/50 | Train Loss: 0.001468 | Val Loss: 0.000575 | Val MSE: 0.000575 | PSNR: 33.52 | SSIM: 0.8088
--> Best model (Task III.B) saved at epoch 22 with validation loss: 0.000575
Validation loss improved from 0.000585 to 0.000575.


Epoch 23/50 | Train Loss: 0.001461 | Val Loss: 0.000589 | Val MSE: 0.000587 | PSNR: 33.47 | SSIM: 0.8032
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 24/50 | Train Loss: 0.001455 | Val Loss: 0.000571 | Val MSE: 0.000569 | PSNR: 33.59 | SSIM: 0.8094
--> Best model (Task III.B) saved at epoch 24 with validation loss: 0.000571
Validation loss improved from 0.000575 to 0.000571.


Epoch 25/50 | Train Loss: 0.001478 | Val Loss: 0.000605 | Val MSE: 0.000606 | PSNR: 33.32 | SSIM: 0.7951
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 26/50 | Train Loss: 0.001443 | Val Loss: 0.000564 | Val MSE: 0.000564 | PSNR: 33.59 | SSIM: 0.8097
--> Best model (Task III.B) saved at epoch 26 with validation loss: 0.000564
Validation loss improved from 0.000571 to 0.000564.


Epoch 27/50 | Train Loss: 0.001437 | Val Loss: 0.000548 | Val MSE: 0.000550 | PSNR: 33.72 | SSIM: 0.8156
--> Best model (Task III.B) saved at epoch 27 with validation loss: 0.000548
Validation loss improved from 0.000564 to 0.000548.


Epoch 28/50 | Train Loss: 0.001439 | Val Loss: 0.000523 | Val MSE: 0.000525 | PSNR: 33.86 | SSIM: 0.8252
--> Best model (Task III.B) saved at epoch 28 with validation loss: 0.000523
Validation loss improved from 0.000548 to 0.000523.


Epoch 29/50 | Train Loss: 0.001437 | Val Loss: 0.000541 | Val MSE: 0.000543 | PSNR: 33.72 | SSIM: 0.8177
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 30/50 | Train Loss: 0.001422 | Val Loss: 0.000545 | Val MSE: 0.000547 | PSNR: 33.70 | SSIM: 0.8146
Validation loss did not improve. EarlyStopping counter: 2/5


Epoch 31/50 | Train Loss: 0.001417 | Val Loss: 0.000530 | Val MSE: 0.000528 | PSNR: 33.85 | SSIM: 0.8209
Validation loss did not improve. EarlyStopping counter: 3/5


Epoch 32/50 | Train Loss: 0.001415 | Val Loss: 0.000544 | Val MSE: 0.000545 | PSNR: 33.73 | SSIM: 0.8149
Validation loss did not improve. EarlyStopping counter: 4/5


Epoch 33/50 | Train Loss: 0.001411 | Val Loss: 0.000513 | Val MSE: 0.000512 | PSNR: 33.93 | SSIM: 0.8284
--> Best model (Task III.B) saved at epoch 33 with validation loss: 0.000513
Validation loss improved from 0.000523 to 0.000513.


Epoch 34/50 | Train Loss: 0.001410 | Val Loss: 0.000510 | Val MSE: 0.000513 | PSNR: 33.94 | SSIM: 0.8285
--> Best model (Task III.B) saved at epoch 34 with validation loss: 0.000510
Validation loss improved from 0.000513 to 0.000510.


Epoch 35/50 | Train Loss: 0.001402 | Val Loss: 0.000532 | Val MSE: 0.000532 | PSNR: 33.81 | SSIM: 0.8191
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 36/50 | Train Loss: 0.001402 | Val Loss: 0.000516 | Val MSE: 0.000513 | PSNR: 33.96 | SSIM: 0.8252
Validation loss did not improve. EarlyStopping counter: 2/5


Epoch 37/50 | Train Loss: 0.001392 | Val Loss: 0.000531 | Val MSE: 0.000532 | PSNR: 33.81 | SSIM: 0.8185
Validation loss did not improve. EarlyStopping counter: 3/5


Epoch 38/50 | Train Loss: 0.001388 | Val Loss: 0.000510 | Val MSE: 0.000509 | PSNR: 34.01 | SSIM: 0.8258
Validation loss did not improve. EarlyStopping counter: 4/5


Epoch 39/50 | Train Loss: 0.001386 | Val Loss: 0.000482 | Val MSE: 0.000482 | PSNR: 34.22 | SSIM: 0.8397
--> Best model (Task III.B) saved at epoch 39 with validation loss: 0.000482
Validation loss improved from 0.000510 to 0.000482.


Epoch 40/50 | Train Loss: 0.001385 | Val Loss: 0.000489 | Val MSE: 0.000491 | PSNR: 34.13 | SSIM: 0.8346
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 41/50 | Train Loss: 0.001377 | Val Loss: 0.000501 | Val MSE: 0.000501 | PSNR: 34.06 | SSIM: 0.8297
Validation loss did not improve. EarlyStopping counter: 2/5


Epoch 42/50 | Train Loss: 0.001374 | Val Loss: 0.000474 | Val MSE: 0.000472 | PSNR: 34.27 | SSIM: 0.8435
--> Best model (Task III.B) saved at epoch 42 with validation loss: 0.000474
Validation loss improved from 0.000482 to 0.000474.


Epoch 43/50 | Train Loss: 0.001374 | Val Loss: 0.000481 | Val MSE: 0.000481 | PSNR: 34.22 | SSIM: 0.8398
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 44/50 | Train Loss: 0.001370 | Val Loss: 0.000514 | Val MSE: 0.000514 | PSNR: 33.93 | SSIM: 0.8258
Validation loss did not improve. EarlyStopping counter: 2/5


Epoch 45/50 | Train Loss: 0.001366 | Val Loss: 0.000477 | Val MSE: 0.000474 | PSNR: 34.24 | SSIM: 0.8413
Validation loss did not improve. EarlyStopping counter: 3/5


Epoch 46/50 | Train Loss: 0.001368 | Val Loss: 0.000462 | Val MSE: 0.000462 | PSNR: 34.36 | SSIM: 0.8476
--> Best model (Task III.B) saved at epoch 46 with validation loss: 0.000462
Validation loss improved from 0.000474 to 0.000462.


Epoch 47/50 | Train Loss: 0.001356 | Val Loss: 0.000488 | Val MSE: 0.000490 | PSNR: 34.11 | SSIM: 0.8342
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 48/50 | Train Loss: 0.001358 | Val Loss: 0.000461 | Val MSE: 0.000458 | PSNR: 34.36 | SSIM: 0.8469
--> Best model (Task III.B) saved at epoch 48 with validation loss: 0.000461
Validation loss improved from 0.000462 to 0.000461.


Epoch 49/50 | Train Loss: 0.001361 | Val Loss: 0.000493 | Val MSE: 0.000490 | PSNR: 34.14 | SSIM: 0.8323
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 50/50 | Train Loss: 0.001355 | Val Loss: 0.000468 | Val MSE: 0.000468 | PSNR: 34.27 | SSIM: 0.8424
Validation loss did not improve. EarlyStopping counter: 2/5
Fine-tuning complete. Best model saved as 'best_model_task3b.pth'.


# Task III.B: Super-Resolution on Real Data with Data Augmentation

Fine tuning our pre-trained super-resolution model from Task III.A using a limited dataset of real HR/LR pairs (300 pairs). To help the model generalize better given the small dataset, I applied data augmentation (e.g., random horizontal flips) during training. I reused previously defined classes and functions (`SRCNN`, `SuperResolutionModel`, `EarlyStopping`, and `evaluate`), and only adjusted the data loading pipeline and training parameters as needed.

The fine-tuning was monitored using validation metrics (MSE, PSNR, SSIM), and early stopping was used to avoid overfitting. The best model is saved for later analysis.


In [35]:
class AugmentedSuperResolutionDataset(SuperResolutionDataset):
    def __init__(self, hr_dir, lr_dir, transform=None):
        super().__init__(hr_dir, lr_dir, transform)
    
    def __getitem__(self, idx):
        sample = super().__getitem__(idx)
        # Apply a simple data augmentation: random horizontal flip
        if np.random.rand() > 0.5:
            sample['lr'] = torch.flip(sample['lr'], dims=[2])
            sample['hr'] = torch.flip(sample['hr'], dims=[2])
        return sample

In [36]:
task3b_hr_dir = '/kaggle/input/specific-task-3b/Dataset 3B/Dataset/HR'
task3b_lr_dir = '/kaggle/input/specific-task-3b/Dataset 3B/Dataset/LR'

In [37]:
dataset_task3b = AugmentedSuperResolutionDataset(task3b_hr_dir, task3b_lr_dir)
train_size = int(0.9 * len(dataset_task3b))
val_size = len(dataset_task3b) - train_size
train_dataset, val_dataset = torch.utils.data.random_split(dataset_task3b, [train_size, val_size])
train_loader = DataLoader(train_dataset, batch_size=8, shuffle=True, num_workers=2)
val_loader = DataLoader(val_dataset, batch_size=8, shuffle=False, num_workers=2)

print(f"Task III.B - Total samples: {len(dataset_task3b)} | Train samples: {len(train_dataset)} | Val samples: {len(val_dataset)}")

Task III.B - Total samples: 300 | Train samples: 270 | Val samples: 30


In [38]:
model_ft = SuperResolutionModel(scale_factor=2).cuda()
pretrained_path = 'best_model.pth'
if os.path.exists(pretrained_path):
    model_ft.load_state_dict(torch.load(pretrained_path))
    print("Pre-trained weights loaded from Task III.A.")
else:
    print("Pre-trained model not found. Please ensure 'best_model.pth' is available.")

# Set a lower learning rate for fine-tuning
optimizer = optim.Adam(model_ft.parameters(), lr=1e-5)
criterion = nn.MSELoss()

# Early stopping and best model saving (reuse EarlyStopping)
early_stopping = EarlyStopping(patience=5, verbose=True)
best_val_loss = float('inf')

Pre-trained weights loaded from Task III.A.


<ipython-input-38-c3b279ca6815>:4: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  model_ft.load_state_dict(torch.load(pretrained_path))


In [39]:
num_epochs = 50  # Maximum epochs for fine-tuning
print("Starting fine-tuning on Task III.B dataset...\n")
for epoch in range(num_epochs):
    model_ft.train()
    running_loss = 0.0
    train_bar = tqdm(train_loader, desc=f"Epoch {epoch+1}/{num_epochs} [Training]", leave=False)
    for batch in train_bar:
        lr = batch['lr'].cuda()
        hr = batch['hr'].cuda()
        optimizer.zero_grad()
        sr = model_ft(lr)
        loss = criterion(sr, hr)
        loss.backward()
        optimizer.step()
        running_loss += loss.item() * lr.size(0)
        train_bar.set_postfix(loss=loss.item())
    
    train_loss = running_loss / len(train_loader.dataset)
    
    # Validation loop
    model_ft.eval()
    val_running_loss = 0.0
    with torch.no_grad():
        val_bar = tqdm(val_loader, desc=f"Epoch {epoch+1}/{num_epochs} [Validation]", leave=False)
        for batch in val_bar:
            lr = batch['lr'].cuda()
            hr = batch['hr'].cuda()
            sr = model_ft(lr)
            loss = criterion(sr, hr)
            val_running_loss += loss.item() * lr.size(0)
            val_bar.set_postfix(loss=loss.item())
    val_loss = val_running_loss / len(val_loader.dataset)
    
    # Compute additional metrics on the validation set
    val_mse, val_psnr, val_ssim = evaluate(model_ft, val_loader)
    
    print(f"Epoch {epoch+1}/{num_epochs} | Train Loss: {train_loss:.6f} | Val Loss: {val_loss:.6f} | "
          f"Val MSE: {val_mse:.6f} | PSNR: {val_psnr:.2f} | SSIM: {val_ssim:.4f}")
    
    # Save the best model for Task III.B
    if val_loss < best_val_loss:
        best_val_loss = val_loss
        torch.save(model_ft.state_dict(), 'best_model_task3b.pth')
        print(f"--> Best model (Task III.B) saved at epoch {epoch+1} with validation loss: {val_loss:.6f}")
    
    early_stopping(val_loss)
    if early_stopping.early_stop:
        print("Early stopping triggered. Ending fine-tuning.")
        break

print("Fine-tuning complete. Best model saved as 'best_model_task3b.pth'.")

Starting fine-tuning on Task III.B dataset...



Epoch 1/50 | Train Loss: 0.006198 | Val Loss: 0.003043 | Val MSE: 0.003040 | PSNR: 28.24 | SSIM: 0.6068
--> Best model (Task III.B) saved at epoch 1 with validation loss: 0.003043


Epoch 2/50 | Train Loss: 0.002879 | Val Loss: 0.001395 | Val MSE: 0.001396 | PSNR: 30.76 | SSIM: 0.5707
--> Best model (Task III.B) saved at epoch 2 with validation loss: 0.001395
Validation loss improved from 0.003043 to 0.001395.


Epoch 3/50 | Train Loss: 0.001977 | Val Loss: 0.001133 | Val MSE: 0.001140 | PSNR: 31.25 | SSIM: 0.5338
--> Best model (Task III.B) saved at epoch 3 with validation loss: 0.001133
Validation loss improved from 0.001395 to 0.001133.


Epoch 4/50 | Train Loss: 0.001801 | Val Loss: 0.001040 | Val MSE: 0.001042 | PSNR: 31.60 | SSIM: 0.5635
--> Best model (Task III.B) saved at epoch 4 with validation loss: 0.001040
Validation loss improved from 0.001133 to 0.001040.


Epoch 5/50 | Train Loss: 0.001720 | Val Loss: 0.000974 | Val MSE: 0.000976 | PSNR: 31.95 | SSIM: 0.6288
--> Best model (Task III.B) saved at epoch 5 with validation loss: 0.000974
Validation loss improved from 0.001040 to 0.000974.


Epoch 6/50 | Train Loss: 0.001661 | Val Loss: 0.000947 | Val MSE: 0.000947 | PSNR: 32.21 | SSIM: 0.6814
--> Best model (Task III.B) saved at epoch 6 with validation loss: 0.000947
Validation loss improved from 0.000974 to 0.000947.


Epoch 7/50 | Train Loss: 0.001624 | Val Loss: 0.000930 | Val MSE: 0.000930 | PSNR: 32.24 | SSIM: 0.6981
--> Best model (Task III.B) saved at epoch 7 with validation loss: 0.000930
Validation loss improved from 0.000947 to 0.000930.


Epoch 8/50 | Train Loss: 0.001613 | Val Loss: 0.000917 | Val MSE: 0.000918 | PSNR: 32.36 | SSIM: 0.7285
--> Best model (Task III.B) saved at epoch 8 with validation loss: 0.000917
Validation loss improved from 0.000930 to 0.000917.


Epoch 9/50 | Train Loss: 0.001590 | Val Loss: 0.000908 | Val MSE: 0.000905 | PSNR: 32.42 | SSIM: 0.7467
--> Best model (Task III.B) saved at epoch 9 with validation loss: 0.000908
Validation loss improved from 0.000917 to 0.000908.


Epoch 10/50 | Train Loss: 0.001570 | Val Loss: 0.000894 | Val MSE: 0.000891 | PSNR: 32.53 | SSIM: 0.7610
--> Best model (Task III.B) saved at epoch 10 with validation loss: 0.000894
Validation loss improved from 0.000908 to 0.000894.


Epoch 11/50 | Train Loss: 0.001558 | Val Loss: 0.000881 | Val MSE: 0.000882 | PSNR: 32.62 | SSIM: 0.7702
--> Best model (Task III.B) saved at epoch 11 with validation loss: 0.000881
Validation loss improved from 0.000894 to 0.000881.


Epoch 12/50 | Train Loss: 0.001550 | Val Loss: 0.000870 | Val MSE: 0.000874 | PSNR: 32.66 | SSIM: 0.7758
--> Best model (Task III.B) saved at epoch 12 with validation loss: 0.000870
Validation loss improved from 0.000881 to 0.000870.


Epoch 13/50 | Train Loss: 0.001534 | Val Loss: 0.000859 | Val MSE: 0.000860 | PSNR: 32.73 | SSIM: 0.7810
--> Best model (Task III.B) saved at epoch 13 with validation loss: 0.000859
Validation loss improved from 0.000870 to 0.000859.


Epoch 14/50 | Train Loss: 0.001521 | Val Loss: 0.000849 | Val MSE: 0.000847 | PSNR: 32.81 | SSIM: 0.7870
--> Best model (Task III.B) saved at epoch 14 with validation loss: 0.000849
Validation loss improved from 0.000859 to 0.000849.


Epoch 15/50 | Train Loss: 0.001511 | Val Loss: 0.000837 | Val MSE: 0.000840 | PSNR: 32.86 | SSIM: 0.7884
--> Best model (Task III.B) saved at epoch 15 with validation loss: 0.000837
Validation loss improved from 0.000849 to 0.000837.


Epoch 16/50 | Train Loss: 0.001496 | Val Loss: 0.000827 | Val MSE: 0.000826 | PSNR: 32.94 | SSIM: 0.7955
--> Best model (Task III.B) saved at epoch 16 with validation loss: 0.000827
Validation loss improved from 0.000837 to 0.000827.


Epoch 17/50 | Train Loss: 0.001481 | Val Loss: 0.000820 | Val MSE: 0.000818 | PSNR: 32.95 | SSIM: 0.7898
--> Best model (Task III.B) saved at epoch 17 with validation loss: 0.000820
Validation loss improved from 0.000827 to 0.000820.


Epoch 18/50 | Train Loss: 0.001472 | Val Loss: 0.000812 | Val MSE: 0.000811 | PSNR: 33.00 | SSIM: 0.7949
--> Best model (Task III.B) saved at epoch 18 with validation loss: 0.000812
Validation loss improved from 0.000820 to 0.000812.


Epoch 19/50 | Train Loss: 0.001462 | Val Loss: 0.000796 | Val MSE: 0.000799 | PSNR: 33.10 | SSIM: 0.8028
--> Best model (Task III.B) saved at epoch 19 with validation loss: 0.000796
Validation loss improved from 0.000812 to 0.000796.


Epoch 20/50 | Train Loss: 0.001454 | Val Loss: 0.000787 | Val MSE: 0.000788 | PSNR: 33.16 | SSIM: 0.8025
--> Best model (Task III.B) saved at epoch 20 with validation loss: 0.000787
Validation loss improved from 0.000796 to 0.000787.


Epoch 21/50 | Train Loss: 0.001448 | Val Loss: 0.000783 | Val MSE: 0.000783 | PSNR: 33.17 | SSIM: 0.7987
--> Best model (Task III.B) saved at epoch 21 with validation loss: 0.000783
Validation loss improved from 0.000787 to 0.000783.


Epoch 22/50 | Train Loss: 0.001442 | Val Loss: 0.000774 | Val MSE: 0.000773 | PSNR: 33.29 | SSIM: 0.8104
--> Best model (Task III.B) saved at epoch 22 with validation loss: 0.000774
Validation loss improved from 0.000783 to 0.000774.


Epoch 23/50 | Train Loss: 0.001439 | Val Loss: 0.000776 | Val MSE: 0.000774 | PSNR: 33.24 | SSIM: 0.8023
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 24/50 | Train Loss: 0.001420 | Val Loss: 0.000767 | Val MSE: 0.000768 | PSNR: 33.33 | SSIM: 0.8127
--> Best model (Task III.B) saved at epoch 24 with validation loss: 0.000767
Validation loss improved from 0.000774 to 0.000767.


Epoch 25/50 | Train Loss: 0.001414 | Val Loss: 0.000763 | Val MSE: 0.000765 | PSNR: 33.31 | SSIM: 0.8090
--> Best model (Task III.B) saved at epoch 25 with validation loss: 0.000763
Validation loss improved from 0.000767 to 0.000763.


Epoch 26/50 | Train Loss: 0.001415 | Val Loss: 0.000757 | Val MSE: 0.000758 | PSNR: 33.37 | SSIM: 0.8084
--> Best model (Task III.B) saved at epoch 26 with validation loss: 0.000757
Validation loss improved from 0.000763 to 0.000757.


Epoch 27/50 | Train Loss: 0.001402 | Val Loss: 0.000759 | Val MSE: 0.000757 | PSNR: 33.40 | SSIM: 0.8094
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 28/50 | Train Loss: 0.001398 | Val Loss: 0.000749 | Val MSE: 0.000751 | PSNR: 33.42 | SSIM: 0.8114
--> Best model (Task III.B) saved at epoch 28 with validation loss: 0.000749
Validation loss improved from 0.000757 to 0.000749.


Epoch 29/50 | Train Loss: 0.001395 | Val Loss: 0.000753 | Val MSE: 0.000755 | PSNR: 33.41 | SSIM: 0.8130
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 30/50 | Train Loss: 0.001387 | Val Loss: 0.000749 | Val MSE: 0.000749 | PSNR: 33.46 | SSIM: 0.8067
--> Best model (Task III.B) saved at epoch 30 with validation loss: 0.000749
Validation loss improved from 0.000749 to 0.000749.


Epoch 31/50 | Train Loss: 0.001386 | Val Loss: 0.000751 | Val MSE: 0.000751 | PSNR: 33.51 | SSIM: 0.8212
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 32/50 | Train Loss: 0.001384 | Val Loss: 0.000746 | Val MSE: 0.000746 | PSNR: 33.48 | SSIM: 0.8129
--> Best model (Task III.B) saved at epoch 32 with validation loss: 0.000746
Validation loss improved from 0.000749 to 0.000746.


Epoch 33/50 | Train Loss: 0.001391 | Val Loss: 0.000751 | Val MSE: 0.000750 | PSNR: 33.55 | SSIM: 0.8253
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 34/50 | Train Loss: 0.001383 | Val Loss: 0.000735 | Val MSE: 0.000738 | PSNR: 33.55 | SSIM: 0.8148
--> Best model (Task III.B) saved at epoch 34 with validation loss: 0.000735
Validation loss improved from 0.000746 to 0.000735.


Epoch 35/50 | Train Loss: 0.001368 | Val Loss: 0.000742 | Val MSE: 0.000742 | PSNR: 33.58 | SSIM: 0.8224
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 36/50 | Train Loss: 0.001371 | Val Loss: 0.000738 | Val MSE: 0.000739 | PSNR: 33.55 | SSIM: 0.8166
Validation loss did not improve. EarlyStopping counter: 2/5


Epoch 37/50 | Train Loss: 0.001376 | Val Loss: 0.000735 | Val MSE: 0.000734 | PSNR: 33.56 | SSIM: 0.8095
--> Best model (Task III.B) saved at epoch 37 with validation loss: 0.000735
Validation loss improved from 0.000735 to 0.000735.


Epoch 38/50 | Train Loss: 0.001363 | Val Loss: 0.000734 | Val MSE: 0.000733 | PSNR: 33.61 | SSIM: 0.8198
--> Best model (Task III.B) saved at epoch 38 with validation loss: 0.000734
Validation loss improved from 0.000735 to 0.000734.


Epoch 39/50 | Train Loss: 0.001357 | Val Loss: 0.000726 | Val MSE: 0.000728 | PSNR: 33.65 | SSIM: 0.8216
--> Best model (Task III.B) saved at epoch 39 with validation loss: 0.000726
Validation loss improved from 0.000734 to 0.000726.


Epoch 40/50 | Train Loss: 0.001356 | Val Loss: 0.000730 | Val MSE: 0.000730 | PSNR: 33.55 | SSIM: 0.8089
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 41/50 | Train Loss: 0.001355 | Val Loss: 0.000725 | Val MSE: 0.000725 | PSNR: 33.71 | SSIM: 0.8272
--> Best model (Task III.B) saved at epoch 41 with validation loss: 0.000725
Validation loss improved from 0.000726 to 0.000725.


Epoch 42/50 | Train Loss: 0.001349 | Val Loss: 0.000719 | Val MSE: 0.000719 | PSNR: 33.74 | SSIM: 0.8255
--> Best model (Task III.B) saved at epoch 42 with validation loss: 0.000719
Validation loss improved from 0.000725 to 0.000719.


Epoch 43/50 | Train Loss: 0.001345 | Val Loss: 0.000719 | Val MSE: 0.000717 | PSNR: 33.75 | SSIM: 0.8245
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 44/50 | Train Loss: 0.001341 | Val Loss: 0.000717 | Val MSE: 0.000717 | PSNR: 33.74 | SSIM: 0.8246
--> Best model (Task III.B) saved at epoch 44 with validation loss: 0.000717
Validation loss improved from 0.000719 to 0.000717.


Epoch 45/50 | Train Loss: 0.001336 | Val Loss: 0.000712 | Val MSE: 0.000712 | PSNR: 33.82 | SSIM: 0.8324
--> Best model (Task III.B) saved at epoch 45 with validation loss: 0.000712
Validation loss improved from 0.000717 to 0.000712.


Epoch 46/50 | Train Loss: 0.001332 | Val Loss: 0.000712 | Val MSE: 0.000711 | PSNR: 33.78 | SSIM: 0.8225
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 47/50 | Train Loss: 0.001332 | Val Loss: 0.000711 | Val MSE: 0.000705 | PSNR: 33.88 | SSIM: 0.8307
--> Best model (Task III.B) saved at epoch 47 with validation loss: 0.000711
Validation loss improved from 0.000712 to 0.000711.


Epoch 48/50 | Train Loss: 0.001331 | Val Loss: 0.000708 | Val MSE: 0.000705 | PSNR: 33.86 | SSIM: 0.8264
--> Best model (Task III.B) saved at epoch 48 with validation loss: 0.000708
Validation loss improved from 0.000711 to 0.000708.


Epoch 49/50 | Train Loss: 0.001330 | Val Loss: 0.000710 | Val MSE: 0.000705 | PSNR: 33.84 | SSIM: 0.8235
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 50/50 | Train Loss: 0.001323 | Val Loss: 0.000707 | Val MSE: 0.000708 | PSNR: 33.85 | SSIM: 0.8267
--> Best model (Task III.B) saved at epoch 50 with validation loss: 0.000707
Validation loss improved from 0.000708 to 0.000707.
Fine-tuning complete. Best model saved as 'best_model_task3b.pth'.


## RCAN Experiment 

**Objective:**  
Experiment with an alternative architecture, the Residual Channel Attention Network (RCAN), to see if it outperforms the baseline model.

**Implementation Steps:**
1. **Model Definition:**  
   - Implement the RCAN architecture including:
     - Channel Attention (CA) modules.
     - Residual Channel Attention Blocks (RCAB).
     - Residual Groups.
2. **Training:**  
   - Train RCAN using the same dataset and training utilities as before.
   - Compare the evaluation metrics (MSE, PSNR, SSIM) to assess performance.
3. **Analysis:**  
   - Present the results and discuss the benefits or drawbacks compared to the baseline model.


In [40]:
class CALayer(nn.Module):
    def __init__(self, channel, reduction=16):
        super(CALayer, self).__init__()
        self.avg_pool = nn.AdaptiveAvgPool2d(1)
        self.conv_du = nn.Sequential(
            nn.Conv2d(channel, channel // reduction, 1, padding=0, bias=True),
            nn.ReLU(inplace=True),
            nn.Conv2d(channel // reduction, channel, 1, padding=0, bias=True),
            nn.Sigmoid()
        )
        
    def forward(self, x):
        y = self.avg_pool(x)
        y = self.conv_du(y)
        return x * y


In [41]:
class RCAB(nn.Module):
    def __init__(self, channel, kernel_size=3, reduction=16, bias=True, activation=nn.ReLU(True)):
        super(RCAB, self).__init__()
        self.body = nn.Sequential(
            nn.Conv2d(channel, channel, kernel_size, padding=kernel_size//2, bias=bias),
            activation,
            nn.Conv2d(channel, channel, kernel_size, padding=kernel_size//2, bias=bias),
            CALayer(channel, reduction)
        )
        
    def forward(self, x):
        out = self.body(x)
        return out + x

In [42]:
class ResidualGroup(nn.Module):
    def __init__(self, channel, n_RCAB, kernel_size=3, reduction=16, bias=True, activation=nn.ReLU(True)):
        super(ResidualGroup, self).__init__()
        modules_body = [RCAB(channel, kernel_size, reduction, bias, activation) for _ in range(n_RCAB)]
        modules_body.append(nn.Conv2d(channel, channel, kernel_size, padding=kernel_size//2, bias=bias))
        self.body = nn.Sequential(*modules_body)
        
    def forward(self, x):
        res = self.body(x)
        return res + x

In [43]:
class RCAN(nn.Module):
    def __init__(self, in_channels=1, out_channels=1, num_features=64, num_rg=5, num_rcab=10, scale=2):
        super(RCAN, self).__init__()
        # Shallow feature extraction
        self.conv1 = nn.Conv2d(in_channels, num_features, kernel_size=3, padding=1, bias=True)
        
        # Residual Groups
        self.residual_groups = nn.Sequential(
            *[ResidualGroup(num_features, num_rcab, kernel_size=3, reduction=16) for _ in range(num_rg)]
        )
        
        # Reconstruction
        self.conv2 = nn.Conv2d(num_features, num_features, kernel_size=3, padding=1, bias=True)
        # Upscaling layers
        self.upscale = nn.Sequential(
            nn.Conv2d(num_features, num_features * (scale ** 2), kernel_size=3, padding=1, bias=True),
            nn.PixelShuffle(scale),
            nn.Conv2d(num_features, out_channels, kernel_size=3, padding=1, bias=True)
        )
    
    def forward(self, x):
        # Shallow feature extraction
        x = self.conv1(x)
        # Deep feature extraction through residual groups
        x = self.residual_groups(x)
        x = self.conv2(x)
        # Upscale to high resolution
        out = self.upscale(x)
        return out

In [44]:
model_rcan = RCAN(in_channels=1, out_channels=1, num_features=64, num_rg=5, num_rcab=10, scale=2).cuda()
print(model_rcan)

RCAN(
  (conv1): Conv2d(1, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (residual_groups): Sequential(
    (0): ResidualGroup(
      (body): Sequential(
        (0): RCAB(
          (body): Sequential(
            (0): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
            (1): ReLU(inplace=True)
            (2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
            (3): CALayer(
              (avg_pool): AdaptiveAvgPool2d(output_size=1)
              (conv_du): Sequential(
                (0): Conv2d(64, 4, kernel_size=(1, 1), stride=(1, 1))
                (1): ReLU(inplace=True)
                (2): Conv2d(4, 64, kernel_size=(1, 1), stride=(1, 1))
                (3): Sigmoid()
              )
            )
          )
        )
        (1): RCAB(
          (body): Sequential(
            (0): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
            (1): ReLU(inplace=True)
            (2): Conv2d(64, 64,

In [45]:
optimizer_rcan = torch.optim.Adam(model_rcan.parameters(), lr=1e-4)
criterion = nn.MSELoss()
early_stopping_rcan = EarlyStopping(patience=5, verbose=True)
best_val_loss_rcan = float('inf')

In [46]:
num_epochs = 50

print("Starting RCAN training...\n")
for epoch in range(num_epochs):
    model_rcan.train()
    running_loss = 0.0
    for batch in tqdm(train_loader, desc=f"Epoch {epoch+1}/{num_epochs} [Training]"):
        lr = batch['lr'].cuda()
        hr = batch['hr'].cuda()
        optimizer_rcan.zero_grad()
        sr = model_rcan(lr)
        loss = criterion(sr, hr)
        loss.backward()
        optimizer_rcan.step()
        running_loss += loss.item() * lr.size(0)
    
    train_loss = running_loss / len(train_loader.dataset)
    
    # Validation
    model_rcan.eval()
    val_running_loss = 0.0
    with torch.no_grad():
        for batch in tqdm(val_loader, desc=f"Epoch {epoch+1}/{num_epochs} [Validation]"):
            lr = batch['lr'].cuda()
            hr = batch['hr'].cuda()
            sr = model_rcan(lr)
            loss = criterion(sr, hr)
            val_running_loss += loss.item() * lr.size(0)
    val_loss = val_running_loss / len(val_loader.dataset)
    
    val_mse, val_psnr, val_ssim = evaluate(model_rcan, val_loader)
    print(f"Epoch {epoch+1}/{num_epochs} | Train Loss: {train_loss:.6f} | Val Loss: {val_loss:.6f} | "
          f"Val MSE: {val_mse:.6f} | PSNR: {val_psnr:.2f} | SSIM: {val_ssim:.4f}")
    
    # Save best model
    if val_loss < best_val_loss_rcan:
        best_val_loss_rcan = val_loss
        torch.save(model_rcan.state_dict(), 'best_model_rcan.pth')
        print(f"--> Best RCAN model saved at epoch {epoch+1} with validation loss: {val_loss:.6f}")
    
    early_stopping_rcan(val_loss)
    if early_stopping_rcan.early_stop:
        print("Early stopping triggered for RCAN. Ending training.")
        break

print("RCAN training complete. Best model saved as 'best_model_rcan.pth'.")

Starting RCAN training...



Epoch 1/50 [Validation]: 100%|██████████| 4/4 [00:00<00:00, 17.52it/s]


Epoch 1/50 | Train Loss: 0.002670 | Val Loss: 0.001109 | Val MSE: 0.001110 | PSNR: 30.84 | SSIM: 0.7139
--> Best RCAN model saved at epoch 1 with validation loss: 0.001109


Epoch 2/50 [Validation]: 100%|██████████| 4/4 [00:00<00:00, 18.05it/s]


Epoch 2/50 | Train Loss: 0.001531 | Val Loss: 0.001081 | Val MSE: 0.001080 | PSNR: 32.26 | SSIM: 0.8262
--> Best RCAN model saved at epoch 2 with validation loss: 0.001081
Validation loss improved from 0.001109 to 0.001081.


Epoch 3/50 [Validation]: 100%|██████████| 4/4 [00:00<00:00, 16.83it/s]


Epoch 3/50 | Train Loss: 0.001574 | Val Loss: 0.001036 | Val MSE: 0.001031 | PSNR: 32.49 | SSIM: 0.8104
--> Best RCAN model saved at epoch 3 with validation loss: 0.001036
Validation loss improved from 0.001081 to 0.001036.


Epoch 4/50 [Validation]: 100%|██████████| 4/4 [00:00<00:00, 17.31it/s]


Epoch 4/50 | Train Loss: 0.001510 | Val Loss: 0.000792 | Val MSE: 0.000794 | PSNR: 32.66 | SSIM: 0.8316
--> Best RCAN model saved at epoch 4 with validation loss: 0.000792
Validation loss improved from 0.001036 to 0.000792.


Epoch 5/50 [Validation]: 100%|██████████| 4/4 [00:00<00:00, 17.25it/s]


Epoch 5/50 | Train Loss: 0.001465 | Val Loss: 0.000816 | Val MSE: 0.000818 | PSNR: 33.52 | SSIM: 0.7882
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 6/50 [Validation]: 100%|██████████| 4/4 [00:00<00:00, 18.19it/s]


Epoch 6/50 | Train Loss: 0.001367 | Val Loss: 0.000870 | Val MSE: 0.000868 | PSNR: 32.87 | SSIM: 0.7740
Validation loss did not improve. EarlyStopping counter: 2/5


Epoch 7/50 [Validation]: 100%|██████████| 4/4 [00:00<00:00, 17.39it/s]


Epoch 7/50 | Train Loss: 0.001302 | Val Loss: 0.000743 | Val MSE: 0.000743 | PSNR: 34.20 | SSIM: 0.8478
--> Best RCAN model saved at epoch 7 with validation loss: 0.000743
Validation loss improved from 0.000792 to 0.000743.


Epoch 8/50 [Validation]: 100%|██████████| 4/4 [00:00<00:00, 17.70it/s]


Epoch 8/50 | Train Loss: 0.001264 | Val Loss: 0.000777 | Val MSE: 0.000761 | PSNR: 33.29 | SSIM: 0.8242
Validation loss did not improve. EarlyStopping counter: 1/5


Epoch 9/50 [Validation]: 100%|██████████| 4/4 [00:00<00:00, 17.03it/s]


Epoch 9/50 | Train Loss: 0.001303 | Val Loss: 0.000834 | Val MSE: 0.000827 | PSNR: 32.93 | SSIM: 0.7566
Validation loss did not improve. EarlyStopping counter: 2/5


Epoch 10/50 [Validation]: 100%|██████████| 4/4 [00:00<00:00, 16.76it/s]


Epoch 10/50 | Train Loss: 0.001291 | Val Loss: 0.000859 | Val MSE: 0.000854 | PSNR: 34.14 | SSIM: 0.8240
Validation loss did not improve. EarlyStopping counter: 3/5


Epoch 11/50 [Validation]: 100%|██████████| 4/4 [00:00<00:00, 16.84it/s]


Epoch 11/50 | Train Loss: 0.001268 | Val Loss: 0.000743 | Val MSE: 0.000744 | PSNR: 33.50 | SSIM: 0.7853
Validation loss did not improve. EarlyStopping counter: 4/5


Epoch 12/50 [Validation]: 100%|██████████| 4/4 [00:00<00:00, 18.26it/s]


Epoch 12/50 | Train Loss: 0.001234 | Val Loss: 0.000748 | Val MSE: 0.000759 | PSNR: 34.39 | SSIM: 0.8452
Validation loss did not improve. EarlyStopping counter: 5/5
Early stopping triggered for RCAN. Ending training.
RCAN training complete. Best model saved as 'best_model_rcan.pth'.


## Conclusion

In this task, we developed and evaluated deep learning-based super-resolution methods for strong lensing images.

### Task III.A: Super-Resolution on Simulated Data
- **Approach:**  
  Implementation baseline SRCNN model that first upsamples low-resolution images using bilinear interpolation and then refines them with convolutional layers.
- **Results:**  
  - **Validation MSE:** ~0.000060  
  - **PSNR:** ~42.27 dB  
  - **SSIM:** ~0.9771  
These metrics indicate excellent reconstruction quality on simulated data, demonstrating that the model effectively captures the key features of strong lensing images.

### Task III.B: Transfer Learning on Real Data
- **Approach:**  
  Utilised pre-trained SRCNN-based model to a limited dataset of real HR/LR pairs (300 images) using transfer learning. The model was fine-tuned with a lower learning rate and evaluated using a 90:10 train–validation split.
- **Results without Data Augmentation:**  
  - **Best Validation Loss:** ~0.000461 (achieved at epoch 48)  
  - **Validation MSE:** ~0.000458  
  - **PSNR:** ~34.36 dB  
  - **SSIM:** ~0.8469  
These results show that the model, when fine-tuned on real data, achieves significant improvements in image resolution. Although the performance is lower than on simulated data, it still demonstrates robust super-resolution capability.

### RCAN Experiment (Alternative Approach)
- **Approach:**  
  I also experimented with a Residual Channel Attention Network (RCAN) to assess whether an alternative architecture could further improve performance. RCAN leverages channel attention mechanisms to emphasize important features.
- **Results:**  
  - **Early Best Model (Epoch 7):**  
    - **Validation Loss:** ~0.000743  
    - **PSNR:** ~34.20 dB  
    - **SSIM:** ~0.8478  
While the RCAN model performed competitively, the differences compared to the SRCNN-based approach were subtle given the limited size of the real dataset.

### Final Remarks
- **Transfer Learning Effectiveness:**  
  The strategy of pre-training on simulated data and then fine-tuning on real data proved effective. The SRCNN-based model achieved excellent results on simulated data and robust performance on real data.
- **Architectural Comparison:**  
  Both the SRCNN-based model and the RCAN approach demonstrated strong performance on the real dataset. Although RCAN introduced additional complexity, its results were comparable to the baseline model.
- **Future Work:**  
  Future work could explore more extensive hyperparameter tuning, additional data augmentation strategies, or hybrid architectures that combine the strengths of both approaches to further enhance super-resolution quality.

Overall, the quantitative metrics (MSE, PSNR, and SSIM) confirm that these deep learning-based super-resolution techniques are effective in enhancing strong lensing images, fulfilling the task requirements.